### PR TITLE
Fetch Asset Observation GraphQL

### PR DESCRIPTION
**Pull request by clairelin135**
Adds `assetObservations` resolver to `GrapheneAssetNode`. Fetches asset observation events from event log storage. Restructures Asset Materialization resolvers, flattening `GrapheneAssetMaterialization`, `GrapheneStepMaterialization`, and `GrapheneMaterialization` into `GrapheneMaterializationEvent`, reusing for `GrapheneObservationEvent`.

Also allows Dagit to handle AssetObservation events, which is currently raising an error when the job run is viewed in Dagit. This PR renders the default event row when asset observation events occur:
image

Requires a followup PR where `GrapheneMaterialization` should be converted to `GrapheneMaterializationEvent` (`GrapheneMaterialization` is still being used for passing materializations in `GrapheneRunStepStats`, which will require a minor restructure of `build_run_step_stats_from_events` to pass the whole event object instead of just the materialization).

### CODE REVIEW

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py [hunk 1]
```diff
@@ -101,6 +102,34 @@ class Meta:
         name = "MaterializationCountByPartition"
 
 
+class GrapheneAssetObservation(graphene.ObjectType):
```
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:105]: Again, what's the difference between ObservationOrMaterialization, ObservationEvent, and AssetObservation?

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py [hunk 2]
```diff
@@ -98,14 +98,26 @@ def get_asset(graphene_info, asset_key):
     return GrapheneAsset(key=asset_key, definition=asset_node)
 
 
-def get_asset_events(graphene_info, asset_key, partitions=None, limit=None, before_timestamp=None):
```
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:101]: prefer that you break this out into `get_materialization_events` / `get_observation_events`.  Do we anticipate a time where we'd fetch both together?  It's fine for this to be broken out as a helper function, but being explicit will reduce future confusion.

**Review by prha — CHANGES_REQUESTED** [review]

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py (hunk 1 above)
Comment (clairelin135) [reply python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:105]: ObservationOrMaterialization is the actual AssetObservation object, so the metadata contained within the asset observation. ObservationEvent contains the event metadata (e.g. which step yielded the observation, the step stats) as well as a reference to the AssetObservation object.

I followed the pattern that we used for AssetMaterialization for AssetObservation, maybe to surface partition and run ID? Not really sure why we had this, but I can modify ObservationEvent to contain the run id and partition so we can remove AssetObservation.

File: python_modules/dagster-graphql/dagster_graphql/schema/logs/events.py [hunk 3]
```diff
@@ -269,12 +269,12 @@ def resolve_metadataEntries(self, _graphene_info):
         return _to_metadata_entries(self.metadata_entries)  # pylint: disable=no-member
 
 
-class GrapheneMaterialization(graphene.ObjectType):
+class GrapheneMaterializationOrObservation(graphene.ObjectType):
```
[the remark this answers is not in the document]
Comment (clairelin135) [reply python_modules/dagster-graphql/dagster_graphql/schema/logs/events.py]: Ok, cool, I will change this to a mixin

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py (hunk 2 above)
Comment (clairelin135) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:101]: I don't anticipate we'd fetch both together without having to filter them into separate lists. I've broken this out into two separate methods.

File: python_modules/dagster-graphql/dagster_graphql/schema/logs/events.py [hunk 4]
```diff
@@ -269,12 +270,12 @@ def resolve_metadataEntries(self, _graphene_info):
         return _to_metadata_entries(self.metadata_entries)  # pylint: disable=no-member
 
 
-class GrapheneMaterialization(graphene.ObjectType):
+class AssetOutputMixin:
```
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/schema/logs/events.py:273]: Rather than `AssetOutputMixin`, consider `AssetMetadataMixin`?

**Comment by prha** [comment]
Aside from updating the graphql tests, this looks pretty good!

File: python_modules/dagster/dagster/core/definitions/events.py [hunk 5]
```diff
@@ -268,6 +268,7 @@ class AssetObservation(
         "_AssetObservation",
         [
             ("asset_key", AssetKey),
+            ("description", Optional[str]),
```
Comment (prha) [inline python_modules/dagster/dagster/core/definitions/events.py:271]: maybe these core changes should be part of a separate diff than the frontend/graphql changes?

**Review by prha — APPROVED** [review]

File: python_modules/dagster/dagster/core/definitions/events.py (hunk 5 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/core/definitions/events.py:271]: I'll make description a `@property` for now and create a separate PR to add the description field. Label is a required field to render the log table.

**Pull request merged**